File: 8-1.py
#!/usr/bin/env python

import re

program = [[inst.strip(), False] for inst in open('8.input').readlines()]

ip = 0
acc = 0

instruction_format = re.compile("(?P<inst>acc|nop|jmp) (?P<oper>[+-][0-9]+)")

while True:
  if program[ip][1]:
    print(f"{acc}")
    break

  program[ip][1] = True

  m = instruction_format.match(program[ip][0])
  if m['inst'] == 'nop':
      ip += 1
  elif m['inst'] == 'acc':
      acc += int(m['oper'])
      ip += 1
  elif m['inst'] == 'jmp':
      ip += int(m['oper'])

File: 8-1.rb
#!/usr/bin/env ruby

program = File.read('8.input').lines.map { |inst| [inst.strip, false] }

ip = 0
acc = 0

instruction_format = /(?<inst>acc|nop|jmp) (?<oper>[+-][[:digit:]]+)/

loop do
  if program[ip][1]
    print("#{acc}\n")
    return
  end

  program[ip][1] = true

  instruction_format.match(program[ip][0]) do |m|
    case m['inst']
    when 'nop'
      ip += 1
    when 'acc'
      acc += m['oper'].to_i
      ip += 1
    when 'jmp'
      ip += m['oper'].to_i
    end
  end
end
